### PR TITLE
Fix incorrect pushdown of simplified domain

### PR DESCRIPTION
**Pull request by raunaqmorarka**
For varchar/char types if the remote database is
case insensitive, a simplified domain should not be
pushed down if it is not a discrete set.

### CODE REVIEW

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/PredicatePushdownController.java [hunk 1]
```diff
@@ -47,10 +47,9 @@
             return FULL_PUSHDOWN.apply(session, domain);
         }
 
-        if (domain.getValues().isDiscreteSet()) {
-            return new DomainPushdownResult(
-                    domain.simplify(getDomainCompactionThreshold(session)),
-                    domain);
+        Domain simplifiedDomain = domain.simplify(getDomainCompactionThreshold(session));
+        if (simplifiedDomain.getValues().isDiscreteSet()) {
+            return new DomainPushdownResult(simplifiedDomain, domain);
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/PredicatePushdownController.java:52]: Add code-level comment explaining why.

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java [hunk 2]
```diff
@@ -259,6 +259,24 @@ public void testPredicatePushdown()
...
+        // varchar IN with small compaction threshold
+        assertThat(query(
+                Session.builder(getSession())
+                        .setCatalogSessionProperty("postgresql", "domain_compaction_threshold", "1")
+                        .build(),
+                "SELECT regionkey, nationkey, name FROM nation WHERE name IN ('POLAND', 'ROMANIA', 'VIETNAM')"))
+                .matches("VALUES " +
+                        "(BIGINT '3', BIGINT '19', CAST('ROMANIA' AS varchar(25))), " +
+                        "(BIGINT '2', BIGINT '21', CAST('VIETNAM' AS varchar(25)))")
+                .isNotFullyPushedDown(FilterNode.class);
```
Comment (findepi) [inline plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:278]: add a comment why Filter is expected to be retained

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java [hunk 3]
```diff
@@ -259,6 +259,24 @@ public void testPredicatePushdown()
...
+
+        // varchar IN with small compaction threshold
+        assertThat(query(
+                Session.builder(getSession())
+                        .setCatalogSessionProperty("postgresql", "domain_compaction_threshold", "1")
+                        .build(),
+                "SELECT regionkey, nationkey, name FROM nation WHERE name IN ('POLAND', 'ROMANIA', 'VIETNAM')"))
+                .matches("VALUES " +
+                        "(BIGINT '3', BIGINT '19', CAST('ROMANIA' AS varchar(25))), " +
+                        "(BIGINT '2', BIGINT '21', CAST('VIETNAM' AS varchar(25)))")
```
Comment (findepi) [inline plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:277]: we should have reusable correctness test for this.

of course, we cannot assert pushdown happening or not (depends on connector)

Comment (kokosing) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:277]: Yes. Notice that this problem may also affect MemSQL and MySQL.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/PredicatePushdownController.java [hunk 4]
```diff
@@ -47,10 +47,9 @@
             return FULL_PUSHDOWN.apply(session, domain);
         }
 
-        if (domain.getValues().isDiscreteSet()) {
```
Comment (kokosing) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/PredicatePushdownController.java:50]: I think we can retain this `if`. If `! domain.getValues().isDiscreteSet()` then it surely `! simplifiedDomain.getValues().isDiscreteSet()`.

File: core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java [hunk 5]
```diff
@@ -779,7 +779,7 @@ public boolean equals(Object obj)
         SortedRangeSet other = (SortedRangeSet) obj;
         return hashCode() == other.hashCode() && // compare hash codes because they are cached, so this is cheap and efficient
                 Objects.equals(this.type, other.type) &&
-                Arrays.equals(inclusive, inclusive) &&
```
Comment (tooptoop4) [inline core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java:782]: 😮

Comment (wendigo) [reply core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java:782]: :O

**Comment by raunaqmorarka** [comment]
Unrelated test failure :
```
2021-07-26T12:22:57.1182923Z [ERROR] Tests run: 219, Failures: 1, Errors: 0, Skipped: 8, Time elapsed: 176.468 s <<< FAILURE! - in TestSuite
2021-07-26T12:22:57.1186608Z [ERROR] testCustomMetricsScanFilter(io.trino.plugin.memory.TestMemoryConnectorTest)  Time elapsed: 0.138 s  <<< FAILURE!
2021-07-26T12:22:57.1188334Z org.junit.ComparisonFailure: expected:<LongCount{total=[2000]}> but was:<LongCount{total=[1332]}>
2021-07-26T12:22:57.1190186Z 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
2021-07-26T12:22:57.1192886Z 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
2021-07-26T12:22:57.1196284Z 	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
2021-07-26T12:22:57.1199666Z 	at io.trino.plugin.memory.TestMemoryConnectorTest.testCustomMetricsScanFilter(TestMemoryConnectorTest.java:156)
2021-07-26T12:22:57.1202355Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2021-07-26T12:22:57.1205041Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2021-07-26T12:22:57.1212786Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2021-07-26T12:22:57.1214922Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2021-07-26T12:22:57.1216619Z 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2021-07-26T12:22:57.1218403Z 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
2021-07-26T12:22:57.1219775Z 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
2021-07-26T12:22:57.1221242Z 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
2021-07-26T12:22:57.1222992Z 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
2021-07-26T12:22:57.1224744Z 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
2021-07-26T12:22:57.1226461Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2021-07-26T12:22:57.1228325Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2021-07-26T12:22:57.1229541Z 	at java.base/java.lang.Thread.run(Thread.java:829)
2021-07-26T12:22:57.1230049Z 
2021-07-26T12:22:57.5200637Z [INFO] 
2021-07-26T12:22:57.5201414Z [INFO] Results:
2021-07-26T12:22:57.5201938Z [INFO] 
2021-07-26T12:22:57.5202457Z [ERROR] Failures: 
2021-07-26T12:22:57.5204010Z [ERROR]   TestMemoryConnectorTest.testCustomMetricsScanFilter:156 expected:<LongCount{total=[2000]}> but was:<LongCount{total=[1332]}>
```

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectorTest.java [hunk 6]
```diff
@@ -815,6 +815,27 @@ public void testCaseSensitiveTopNPushdown()
         }
     }
 
+    @Test
+    public void testCaseInsensitivePredicate()
+    {
+        assertThat(query("SELECT address FROM customer WHERE address IN ('01bR7OOM6zPqo29DpAq', 'BJYZYJQk4yD5B', 'a6M1wdC44LW')"))
+                .skippingTypesCheck()
+                .matches("VALUES ('BJYZYJQk4yD5B'), ('a6M1wdC44LW'), ('01bR7OOM6zPqo29DpAq')");
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectorTest.java:823]: nit: redundant braces in VALUES

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectorTest.java [hunk 7]
```diff
@@ -815,6 +815,27 @@ public void testCaseSensitiveTopNPushdown()
...
+
+        assertThat(query("SELECT COUNT(*) FROM customer WHERE address >= '01bR7OOM6zPqo29DpAq' AND address <= 'a6M1wdC44LW'"))
+                .matches("VALUES BIGINT '858'");
+
+        assertThat(query(
+                // Force conversion to a range predicate which will exclude the row corresponding to 'BJYZYJQk4yD5B'
+                // if the range predicate is pushed into a case insensitive connector
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectorTest.java:830]: will -> would
is pushed -> were pushed

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
@raunaqmorarka see failures

**Comment by raunaqmorarka** [comment]
> @raunaqmorarka see failures

Failures were unrelated flakes, I'll keep an eye on the latest run

**Pull request merged**